### PR TITLE
tree: Add PCI physical slot number for controller

### PR DESCRIPTION
**Pull request by usaleem-ix**
This commit introduces a physical slot field for controller, that contains the PCI physcial slot number for controller device.

### CODE REVIEW

**Comment by usaleem-ix** [comment]
In case, there are multiple NVME drives present on the platform, it's hard to identify which NVME drive is present in which slot. The slot number is usually helpful in determining the location. It is cross reference-able from `lspci`, but it would be nice to have a direct option.

Below is the output from nvme-cli, if this gets approved and merged:
```
Device   SN                   MN                                       FR       TxPort Address        Slot   Subsystem    Namespaces      
-------- -------------------- ---------------------------------------- -------- ------ -------------- ------ ------------ ----------------
nvme9    7150A006T1Q8         KCM6DVUL1T60                             0105     pcie   0000:43:00.0   3      nvme-subsys9 nvme9n1
nvme8    X270A01ST1Q8         KCM6DVUL1T60                             0106     pcie   0000:42:00.0   1      nvme-subsys8 nvme8n1
nvme19   X270A01WT1Q8         KCM6DVUL1T60                             0106     pcie   0000:07:00.0   10     nvme-subsys19 nvme19n1
nvme18   2240A008T1Q8         KCM6DVUL1T60                             0106     pcie   0000:06:00.0   8      nvme-subsys18 nvme18n1
nvme17   X270A00ZT1Q8         KCM6DVUL1T60                             0106     pcie   0000:05:00.0   6      nvme-subsys17 nvme17n1
nvme16   2240A00AT1Q8         KCM6DVUL1T60                             0106     pcie   0000:04:00.0   4      nvme-subsys16 nvme16n1
nvme15   2240A00KT1Q8         KCM6DVUL1T60                             0106     pcie   0000:03:00.0   2      nvme-subsys15 nvme15n1
nvme14   S59BNM0R807848Z      Samsung SSD 970 EVO Plus 250GB           2B2QEXM7 pcie   0000:01:00.0   32     nvme-subsys14 nvme14n1
nvme12   2240A00FT1Q8         KCM6DVUL1T60                             0106     pcie   0000:46:00.0   9      nvme-subsys12 nvme12n1
nvme11   X270A00MT1Q8         KCM6DVUL1T60                             0106     pcie   0000:45:00.0   7      nvme-subsys11 nvme11n1
nvme10   X270A015T1Q8         KCM6DVUL1T60                             0106     pcie   0000:44:00.0   5      nvme-subsys10 nvme10n1
```

File: src/nvme/tree.c [hunk 1]
```diff
@@ -1222,6 +1230,48 @@ static char *nvme_ctrl_lookup_subsystem_name(nvme_root_t r,
...
+static char *nvme_ctrl_lookup_phy_slot(nvme_root_t r, const char *address)
+{
+	char *target_addr;
+	char *addr;
+	char *path;
+	int found = 0;
+	int ret;
+	struct dirent *entry;
+
+	DIR *slots_dir = opendir(nvme_slots_sysfs_dir);
```
Comment (igaw) [inline src/nvme/tree.c:1242]: Split the variable declaration and usage to keep it consistent.

```c
DIR *slots_dir;

slots_dir = opendir(...);
```

File: src/nvme/tree.c [hunk 2]
```diff
@@ -1222,6 +1230,48 @@ static char *nvme_ctrl_lookup_subsystem_name(nvme_root_t r,
...
+	if (slots_dir == NULL) {
```
Comment (igaw) [inline src/nvme/tree.c:1243]: `if (!slots_dir) { ...}` is the preferred way to check for NULL pointers.

File: src/nvme/tree.c [hunk 3]
```diff
@@ -1222,6 +1230,48 @@ static char *nvme_ctrl_lookup_subsystem_name(nvme_root_t r,
...
+		nvme_msg(r, LOG_WARNING, "failed to open slots dir %s\n",
+		nvme_slots_sysfs_dir);
+		return NULL;
+	}
+
+	target_addr = strndup(address, 10);
+	while ((entry = readdir(slots_dir)) != NULL) {
```
Comment (igaw) [inline src/nvme/tree.c:1250]: `while (!(entry = readdir))`

File: src/nvme/tree.c [hunk 4]
```diff
@@ -1222,6 +1230,48 @@ static char *nvme_ctrl_lookup_subsystem_name(nvme_root_t r,
...
+		if (entry->d_type == DT_DIR && strncmp(entry->d_name, ".", 1) != 0
+			&& strncmp(entry->d_name, "..", 2) != 0) {
+			ret = asprintf(&path, "/sys/bus/pci/slots/%s", entry->d_name);
```
Comment (igaw) [inline src/nvme/tree.c:1253]: Align the conditions on brackets and have the condition at the end of the line:

```
		if (entry->d_type == DT_DIR && 
		    strncmp(entry->d_name, ".", 1) != 0 &&
		    strncmp(entry->d_name, "..", 2) != 0)  {
```

File: src/libnvme.map [hunk 5]
```diff
@@ -9,6 +9,7 @@ LIBNVME_1_5 {
 		nvme_root_set_application;
 		nvme_subsystem_get_application;
 		nvme_subsystem_set_application;
+		nvme_ctrl_get_phy_slot;
```
Comment (igaw) [inline src/libnvme.map:12]: add in alphabetically the new function

**Review by igaw — CHANGES_REQUESTED** [review]
Alright, got it. I suppose in larger setups this is really helpful.

A couple of formatting nitpicks.

File: src/nvme/tree.c (hunk 1 above)
Comment (usaleem-ix) [reply src/nvme/tree.c]: Updated.

File: src/nvme/tree.c (hunk 2 above)
Comment (usaleem-ix) [reply src/nvme/tree.c]: Updated.

**Review by igaw — APPROVED** [review]

**Comment by igaw** [comment]
Thanks!

**Comment by igaw** [comment]
On my system I don't get the slot info:

```
Device   SN                   MN                                       FR       TxPort Address        Slot   Subsystem    Namespaces      
-------- -------------------- ---------------------------------------- -------- ------ -------------- ------ ------------ ----------------
nvme0    49MA23BZK03N         KXG60ZNV512G NVMe TOSHIBA 512GB          10604107 pcie   0000:3d:00.0          nvme-subsys0 nvme0n1
```

And indeed there is no entry for this m2 nvme device in `/sys/bus/pci/slots`. I suspect the slot only comes into play if it's not a m2 slot, right?

**Comment by hreinecke** [comment]
No, that's a DMI thing. If dmi provides information about that PCI slot you'll get that information.
Try 'dmidecode' and see if it lists the PCI device numbers.

**Comment by usaleem-ix** [comment]
Physical Slot was added to PCIe as part of hotplug specification. So the PCI slots that do not support hotplug, may not show up under `/sys/bus/pci/slots`.

**Comment by igaw** [comment]
Thanks for the info. As suspected it doesn't show up in `dmidecode`.

**Pull request merged**